core/templates/dev/head/pages/exploration_editor/editor_tab/StateSolutionHelperService.js

```diff
@@ -0,0 +1,116 @@
+  'GraphInput',
+  'SetInput',
+  'MathExpressionInput',
+  'ItemSelectionInput',
+  'LogicProof'
+]);
+
+oppia.constant('INTERACTION_OBJECT_TYPES', {
+  CodeRepl: 'CodeString',
+  GraphInput: 'Graph',
+  ImageClickInput: 'ImageWithRegions',
+  PencilCodeEditor: 'CodeString',
+  MathExpressionInput: 'UnicodeString',
+  MusicNotesInput: 'MusicPhrase',
+  NumericInput: 'Real',
+  TextInput: 'NormalizedString',
+  LogicProof: 'LogicQuestion'
+});
+
+oppia.factory('StateSolutionHelperService', [
+  'oppiaHtmlEscaper', 'responsesService', 'stateInteractionIdService',
+  'explorationContextService', 'editorContextService', '$injector',
+  'explorationStatesService', 'angularNameService', 'stateSolutionService',
+  'AnswerClassificationService', 'alertsService',
+  'SUPPORTED_INTERACTIONS', 'INTERACTION_OBJECT_TYPES',
+  function(oppiaHtmlEscaper, responsesService, stateInteractionIdService,
+    explorationStatesService, editorContextService, $injector,
+    explorationContextService, angularNameService, stateSolutionService,
+    AnswerClassificationService, alertsService,
+    SUPPORTED_INTERACTIONS, INTERACTION_OBJECT_TYPES) {
+    return {
+      isSupportedInteraction: function (id) {
+        return (SUPPORTED_INTERACTIONS.indexOf(id) !== -1);
+      },
+      getInteractionObjectType: function (id) {
+        return INTERACTION_OBJECT_TYPES[id];
+      },
+      getCorrectAnswerObject: function(answer, objectType) {
+        if (objectType === 'CodeString') {
+          return {
+            code: answer,
+            output: '',
+            evaluation: '',
+            error: ''
+          }
+        } else {
+          return answer;
+        }
+      },
+      verifyAndSaveAnswer: function(explorationId, state, correctAnswer) {
+        var interactionId = stateInteractionIdService.savedMemento;
+        var rulesServiceName = (
+          angularNameService.getNameOfInteractionRulesService(interactionId));
+        // Inject RulesService dynamically.
+        var rulesService = $injector.get(rulesServiceName);
+        AnswerClassificationService.getMatchingClassificationResult(
+          explorationId, state, correctAnswer, true, rulesService
+        ).then(function(result) {
+          if (editorContextService.getActiveStateName() !== (
+              result.outcome.dest)) {
+            stateSolutionService.saveDisplayedValue();
+            explorationStatesService.getState(
+              editorContextService.getActiveStateName()
+            ).interaction.markSolutionAsValid();
+          } else {
+            alertsService.addInfoMessage('That solution does not lead ' +
+              'to the next state!');
+            explorationContextService.getState(
+              editorContextService.getActiveStateName()
+            ).interaction.markSolutionAsInvalid();
+            stateSolutionService.saveDisplayedValue();
+          }
+        });
+      },
+      verifySolution: function(explorationId, state, correctAnswer) {
```

Comment:
Same thing -- refactor, and reduce the responsibility. Perhaps with that change you'll end up needing just one function anyway.